src/rust/engine/src/externs/engine_aware.rs

```diff
@@ -43,6 +43,46 @@ impl EngineAwareInformation for Message {
   }
 }
 
+pub struct Metadata {}
+
+impl EngineAwareInformation for Metadata {
+  type MaybeOutput = Vec<(String, Value)>;
+
+  fn retrieve(_types: &Types, value: &Value) -> Option<Self::MaybeOutput> {
+    let metadata_val = match externs::call_method(&value, "metadata", &[]) {
+      Ok(value) => value,
+      Err(py_err) => {
+        let failure = Failure::from_py_err(py_err);
+        log::error!("Error calling `metadata` method: {}", failure);
+        return None;
+      }
+    };
+
+    let metadata_val = externs::check_for_python_none(metadata_val)?;
+    let gil = Python::acquire_gil();
+    let py = gil.python();
+
+    let mut output = Vec::new();
+    let metadata_dict: &PyDict = metadata_val.cast_as::<PyDict>(py).ok()?;
+
+    for (key, value) in metadata_dict.items(py).into_iter() {
+      let key_name: String = match key.extract(py) {
+        Ok(s) => s,
+        Err(e) => {
+          log::error!(
+            "Error in EngineAware.metadata() implementation - non-string key: {:?}",
```

Comment:
we should log the key here (if we don't.. since I don't fully understand this rust syntax yet)

Reply:
This logs the Rust representation of the Python FFI API object representing the key.